From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Udvare <audvare@gmail.com>
Date: Fri, 9 Oct 2026 18:59:42 -0400
Subject: [PATCH] Disable constant ObjC literals when the compiler supports it

Newer Clang emits ObjC collection and number literals as compile-time
constant objects (NSConstantArray, NSConstantDictionary,
NSConstantIntegerNumber, __NSArray0__struct) that Darling's Foundation
does not provide, so CFNetwork, Security, iokitd, pam_opendirectory and
others failed to link. Pass -fno-objc-constant-literals when available.

Signed-off-by: Andrew Udvare <audvare@gmail.com>
---
 CMakeLists.txt | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index f7346bc9d..61d074e0d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -120,6 +120,15 @@ endif()
 
 SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${IGNORED_WARNINGS}")
 SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${IGNORED_WARNINGS}")
+
+# newer Clang emits ObjC literals as constant objects (NSConstantArray, etc.) that our Foundation does not provide
+include(CheckCCompilerFlag)
+check_c_compiler_flag(-fno-objc-constant-literals HAS_NO_OBJC_CONSTANT_LITERALS)
+if(HAS_NO_OBJC_CONSTANT_LITERALS)
+	set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-objc-constant-literals")
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-objc-constant-literals")
+endif()
+
 set(CMAKE_ASM_FLAGS "${CMAKE_ASM_FLAGS} ${ASM_IGNORED_WARNINGS}")
 SET(CMAKE_C_FLAGS_DEBUG "-O0 -ggdb")
 SET(CMAKE_CXX_FLAGS_DEBUG "-O0 -ggdb")
